Revamp Magento 2 Module

Xaxxis opened the pull request:
- Refactoring code, handling deprecated code
- Merge all module snap
- Add feature refund online from Magento admin panel
- Handling notification refund to close order
- Add feature can use multiple Midtrans account
- Handling logic payload for digital product
- adding class logger
- add new readme

Commit:
delete app code directory

Delete the app & code directory to new structure

Commit:
Move midtrans php library to Gateway directory

Now midtransphp library merge to module gateway directory

Commit:
Block class to handle view logic frontend

Commit:
up version module to 2.3.0

Commit:
Modify Controller class extend AbstractAction

- Now controller class is extend from AbstractAction, except Pending, Index class
- AbstractAction class to handle action order, like create invoice, getOrder, setOrderStatu
- Notification class can receive multiple midtrans account
- Redirect class is a class to handle get token snap and can use multiple midtrans account

Commit:
UI XML file for module setting in magento dashboard

Now all module snap, specific, installment, offline module setting is merge to one module Midtrans_Snap

Commit:
XML Config default for module

Commit:
Dependency Injection for model class and helper

Commit:
Dependency injection setup for frontend

Commit:
Module routes

Commit:
Gateway class from midtrans-php library

All class in gateway directory is from midtrans php-library

Commit:
Helper class to get config from admin setting

Commit:
License file

Commit:
Handling Logger

Tha main class for logger is MidtransLogger, the class in directory handler is handling category file of log

Commit:
Get Module version for dashboard

Commit:
For status order

Commit:
Model class for payment code

All model is extend AbstractPayment class for handling refund, getPaymentCode

Commit:
Model for get Setting from config

Commit:
Config Provider class

the class handling configuration frontend, like client-key, environment, merchant-id, module version

Commit:
Csrf skip validator

Commit:
All view frontend

Commit:
UpgradeSchema for adding midtrans_trx_id in sales_order table

This class running once when run bin/magento setup:upgrade

Commit:
Update handle logger

Commit:
Change beta docs link

Commit:
Add comment for every snap model

Commit:
deleted unused comment

Commit:
Update readme

rizdaprasetya commented on README.md:
```diff
@@ -1,70 +1,298 @@
+1. Install the plugins: composer require midtrans/snap
+2. Enable the plugin:  bin/magento module:enable Midtrans_Snap
+3. Execute upgrade script : bin/magento setup:upgrade
+4. Flush cache storage :  bin/magento cache:flush
```
please format into code, eg: `composer require midtrans/snap`

Xaxxis replied:
done

rizdaprasetya commented on README.md:
```diff
@@ -1,70 +1,298 @@
+
+
+
+![](images/Midtrans-Magento1.png "image_tooltip")
```
You can upload image by opening Github issue in this repo, upload the image as the issue comment.  Once comment posted and image appear, copy the image url. Paste it to this readme.

Xaxxis replied:
done

rizdaprasetya commented on README.md:
```diff
@@ -1,70 +1,298 @@
+
+4. In the Midtrans - Accept Online Payment section, click Basic Settings and fill out the following fields:
+
+<table>
```
Please use Markdown format table instead of HTML table, to make it easier to change in future.
https://www.markdownguide.org/extended-syntax/#tables

Xaxxis replied:
done

rizdaprasetya commented on README.md:
```diff
@@ -1,70 +1,298 @@
+
+
+
+# How to use plugins
```
let's rename to `Plugin Usage Instruction`

Xaxxis replied:
done

rizdaprasetya commented on README.md:
```diff
@@ -1,70 +1,298 @@
+Note: Access Keys are unique for every merchant. Server Key is secret, please always keep Server Key confidential.
+
+
+5. In the Log Options section, you will get default for request, notification and error log is turn on except Throw exception
```
`Log Options` should be bold. 
> you will get default for request, notification and error log is turn on except Throw exception

Bit unclear, probably change to e.g:
> The default value is `on` for `request`, `notification` and `error log`. Except `Throw Exception`, is off by default.

Xaxxis replied:
done

rizdaprasetya commented on README.md:
```diff
@@ -1,70 +1,298 @@
+
+1. **Snap payment integration**
+
+    This is the default Snap for Midtrans Magento plugins, Snap payment will auto-enabled when install the Midtrans plugins. Midtrans will show the available payment method on the Snap payment screen.
```
> will be auto-enabled

> when installing

rizdaprasetya replied:
@Xaxxis reminder to revise wrong phrase to above suggested phrases.

rizdaprasetya commented on README.md:
```diff
@@ -1,70 +1,298 @@
+    This is the default Snap for Midtrans Magento plugins, Snap payment will auto-enabled when install the Midtrans plugins. Midtrans will show the available payment method on the Snap payment screen.
+
+2. **Specific Payment integration | Optional** \
+Specific payment it’s design as optional for specific payment that specify in Allow payment method field, Midtrans Snap will show directly payment method on the Snap screen.
```
> Optional (additional), specifically use this if merchant wants to customize `Allow payment method` field, Midtrans Snap will show selected payment method on the Snap screen.

rizdaprasetya commented on README.md:
```diff
@@ -1,70 +1,298 @@
+If you need custom configuration you can feel free to change the default configuration with the following fields 
+
+
+<table>
```
same as previous table review

rizdaprasetya commented on README.md:
```diff
@@ -1,70 +1,298 @@
+    Offline payment it’s desing as optional for Installment where the Card Issuer and Acquiring Bank don't have to be same entity (e.g: BNI Card and Mandiri Acquiring Bank)
+
+
+Note: You can use different Midtrans Account for every Snap model payment method, should configure the access-key in Optional section `“Use another Midtrans account”`. If the optional access-key is empty, the plugins will automatically use access key on Basic Settings.
```
> ... access-key is optional on each `Use another Midtrans account` sectuon. If left empty, ... access key from **Basic Settings**.

rizdaprasetya commented on README.md:
```diff
@@ -1,70 +1,298 @@
+
+Note: You can use different Midtrans Account for every Snap model payment method, should configure the access-key in Optional section `“Use another Midtrans account”`. If the optional access-key is empty, the plugins will automatically use access key on Basic Settings.
+
+If you need custom configuration you can feel free to change the default configuration with the following fields 
```
> In case you need to customize configuration these field are configurable, and described as follows:

rizdaprasetya replied:
@Xaxxis reminder to revise wrong phrase to above suggested phrases.

rizdaprasetya commented on README.md:
```diff
@@ -1,70 +1,298 @@
+
+## Magento Midtrans Notifications Handling
+
+Login to your [Midtrans Account](https://dashboard.midtrans.com/settings/vtweb_configuration), select your environment (sandbox/production), go to menu **settings** > **configuration** and feel out the following URL fields:
```
> select to your preferred environment

> fill

rizdaprasetya commented on README.md:
```diff
@@ -1,70 +1,298 @@
+
+    `https://[your-site-url]/snap/payment/notification`
+
+*   Finish Redirect URL:
```
Please add Snap preference settings, like
https://github.com/veritrans/SNAP-Woocommerce/#midtrans-map-configuration

rizdaprasetya commented on README.md:
```diff
@@ -1,70 +1,298 @@
+
+## How to online refund transaction
+
+You can request refunds either in the [Midtrans Dashboard](https://dashboard.midtrans.com/transactions) or in the Magento admin. After a refund is issued, it cannot be cancelled or undone. Before you trigger this request, make sure that the refund amount and any other details are correct. The online refund function is available only for payment method gopay and credit card.
```
> from the

not 
> in the

rizdaprasetya replied:
@Xaxxis reminder to revise wrong phrase to above suggested phrases.

rizdaprasetya commented on README.md:
```diff
@@ -1,70 +1,298 @@
+
+## How to online refund transaction
+
+You can request refunds either in the [Midtrans Dashboard](https://dashboard.midtrans.com/transactions) or in the Magento admin. After a refund is issued, it cannot be cancelled or undone. Before you trigger this request, make sure that the refund amount and any other details are correct. The online refund function is available only for payment method gopay and credit card.
```
> ... refund feature is available for payment ...

rizdaprasetya replied:
@Xaxxis reminder to revise wrong phrase to above suggested phrases.

rizdaprasetya commented on README.md:
```diff
@@ -1,70 +1,298 @@
+
+You can request refunds either in the [Midtrans Dashboard](https://dashboard.midtrans.com/transactions) or in the Magento admin. After a refund is issued, it cannot be cancelled or undone. Before you trigger this request, make sure that the refund amount and any other details are correct. The online refund function is available only for payment method gopay and credit card.
+
+If you make a refund in the Midtrans Dashboard, a Refund notification is sent to Magento, set transaction state to CLOSE and for now is not created the credit memo.
```
> refund

not
> refund

> from the

rizdaprasetya replied:
unclear:
> set transaction state to CLOSE and for now is not created the credit memo.

please clarify. "it will change transaction state to **Close**?

rizdaprasetya replied:
close or closed?

rizdaprasetya replied:
@Xaxxis reminder

rizdaprasetya commented on README.md:
```diff
@@ -1,70 +1,298 @@
+1. Log in to your Magento admin panel. 
+2. In the menu, go to **Sales** > **Orders**. This opens the order overview page. 
+3. Click on the order you want to refund.
+4. In the **Order list View** left-hand navigation sidebar, click **Invoices**tab.
```
this cause invalid syntax, please add space
```
**Invoices** tab.
```

rizdaprasetya replied:
@Xaxxis reminder

rizdaprasetya commented on README.md:
```diff
@@ -1,70 +1,298 @@
+2. In the menu, go to **Sales** > **Orders**. This opens the order overview page. 
+3. Click on the order you want to refund.
+4. In the **Order list View** left-hand navigation sidebar, click **Invoices**tab.
+5. In the invoice list page, selected the order. click the view button on invoice you need to request online refund.
```
please bold the element to click etc.

rizdaprasetya replied:
@Xaxxis

rizdaprasetya commented on README.md:
```diff
@@ -1,70 +1,298 @@
+3. Click on the order you want to refund.
+4. In the **Order list View** left-hand navigation sidebar, click **Invoices**tab.
+5. In the invoice list page, selected the order. click the view button on invoice you need to request online refund.
+6. Click **Credit Memo** on the top-right corner of the page. ****
```
typo?
```
****
```

rizdaprasetya commented on README.md:
```diff
@@ -1,70 +1,298 @@
+5. In the invoice list page, selected the order. click the view button on invoice you need to request online refund.
+6. Click **Credit Memo** on the top-right corner of the page. ****
+7. In the **New Memo for Invoice** page, scroll down to the **Refund Totals** section.
+8. In this section, you can request a online **Refund** or a **Refund Offline**.
```
no need `a`

Xaxxis replied:
done

rizdaprasetya commented on README.md:
```diff
@@ -1,70 +1,298 @@
+6. Click **Credit Memo** on the top-right corner of the page. ****
+7. In the **New Memo for Invoice** page, scroll down to the **Refund Totals** section.
+8. In this section, you can request a online **Refund** or a **Refund Offline**.
+    *   **Refund:** This option request refund Online to Midtrans, Midtrans automatically send refund notification and changes to **Closed**.
```
> This option will

Xaxxis replied:
done

rizdaprasetya commented on README.md:
```diff
@@ -1,70 +1,298 @@
+    *   **Refund:** This option request refund Online to Midtrans, Midtrans automatically send refund notification and changes to **Closed**.
+    *   **Refund Offline**: An offline refund does not trigger request refund to midtrans it’s only refund in Magento side. You need to take action and carry out the refund manually from Midtrans dashboard.
+
+After a refund operation, the order status changes to **Closed**. This change is controlled by the Magento system, and we cannot influence it. The status change does not mean that the refund was carried out successfully on our side. If the transaction status in Midtrans dashboard changes to **REFUND**, then the refund went through successfully
```
> we cannot influence it
who are we? Midtrans? Midtrans Notification?

not `influence` but `change/modify`

> The status change may not mean that the refund has carried out successfully on our side.

> When the transaction status in  ...

Xaxxis replied:
done

rizdaprasetya commented on Plugin/CsrfValidatorSkip.php:
```diff
@@ -0,0 +1,41 @@
+<?php
+/**
+ *
+ *        c.                                            c.  :.
+ *        E1                                            E.  !)
+ * ::.    E1    !3.        ,5"`'\.;F'`"t.   i.   cF'`'=.E.  !7''`   ;7""   '""!.    ;7"'"!.   ;7'`'*=
+ * ::.    E1    !3.        t.    !)     t   i.  tL      t.  !)     !)     ,,...;)  :1     I.  !t.,
+ * ::.    E1    !3.        t.    !)     t   i.  E.      E.  !)     !L    t'    :1  :1     !)    ``"1.
+ * ::.    E1    !3.        t.    !)     t   i.  '1.,  ,ct.  !1,    !L    1.  ,;31  :1     !) -..   ;7
+ * '      E1    `'         `            `   `     ``'`  `    `'``  `      `''`  `   `     `    ``'`
+ *        E7
```
This looks good, but functionally not good to add this on every files. It can over-bloat the code, line numbers and excess file size. Please only use it on main important files, not every files.

Xaxxis replied:
done, removed

rizdaprasetya commented on Plugin/CsrfValidatorSkip.php:
```diff
@@ -0,0 +1,41 @@
+ * '      E1    `'         `            `   `     ``'`  `    `'``  `      `''`  `   `     `    ``'`
+ *        E7
+ *
+ * Midtrans Snap Magento 2 Module
+ *
+ * Copyright (c) 2020 Midtrans PT.
+ * This file is open source and available under the MIT license.
+ * See the LICENSE file for more info.
+ *
+ */
```
No need to include copyright, to encourage people to contribute. No need to state MIT license, as long as the root folder have the License file. We can remove this.

You may however add `author` and your name/contact on header comment of important files if you want.

Xaxxis replied:
done, removed

Commit:
Fix setup version

rizdaprasetya reviewed (changes requested):
This only just review the Readme and some minor format. The functional review should happen when you are doing "demo/sharing session".

Overall good 👍 Mostly just found issue on grammar, typo, etc. Please improve accordingly. Or use grammar check tools like "Grammarly" etc.

No need to create separate google docs for install instruction, as long as all the information is available on Readme (the image is displayed on the readme, etc.). Having google docs may means you need to edit 2 things if there are changes in the future. But it is up to you, if you think google docs is fine to maintain, then it's ok.

Commit:
Update readme from pull request review

Commit:
Update image readme

Commit:
Remove copyright on class

Commit:
Update from meeting review

- drop custom schema for finish landing page BCA Klikpay handling, use server_key from basic setting
- ignore deny status handling on notif for now, just addCommentStatusHistory
- Remove Upgrade schema
- Update readme

Commit:
Update README.md

rizdaprasetya commented on README.md:
```diff
@@ -100,6 +95,8 @@ Specific payment it’s design as optional for specific payment that specify in
 
 >Note: You can use different Midtrans Account for every Snap model payment method, should configure the access-key in Optional section `“Use another Midtrans account”`. If the optional access-key is empty, the plugins will automatically use access key on Basic Settings.
 
+>INFO: BCA Klikpay only can use in Snap payment integration
```
> INFO: The built-in BCA Klikpay landing page for now will only use server key from basic settings of **Snap payment integration**

Xaxxis replied:
Done

rizdaprasetya commented on etc/adminhtml/system/midtrans_snap_installment_payment.xml:
```diff
@@ -0,0 +1,76 @@
+            </field>
+        </group>
+        <group id="installment_access" translate="label" type="text" sortOrder="101" showInDefault="1" showInWebsite="1" showInStore="1">
+            <label><![CDATA[Optional - Use another Midtrans account for Installment payment]]></label>
```
fix:
> Use different Midtrans account for ...

Xaxxis replied:
done

rizdaprasetya commented on etc/adminhtml/system/midtrans_snap_offline_payment.xml:
```diff
@@ -0,0 +1,89 @@
+            </field>
+        </group>
+        <group id="offline_access" translate="label" type="text" sortOrder="101" showInDefault="1" showInWebsite="1" showInStore="1">
+            <label><![CDATA[Optional - Use another Midtrans account for Offline payment]]></label>
```
fix:
> Use different Midtrans account for ...

Xaxxis replied:
done

rizdaprasetya commented on etc/adminhtml/system/midtrans_snap_specific_payment.xml:
```diff
@@ -0,0 +1,85 @@
+            </field>
+        </group>
+        <group id="specific_access" translate="label" type="text" sortOrder="101" showInDefault="1" showInWebsite="1" showInStore="1">
+            <label><![CDATA[Optional - Use another Midtrans account for Specific payment]]></label>
```
fix:
> Use different Midtrans account for ...

Xaxxis replied:
done

rizdaprasetya commented on etc/adminhtml/system/midtrans_snap_installment_payment.xml:
```diff
@@ -0,0 +1,76 @@
+    <group id="installment" translate="label" type="text" sortOrder="50" showInDefault="1" showInWebsite="1" showInStore="1">
+        <label><![CDATA[Optional - Snap Online Installment Payment Integration]]></label>
+        <frontend_model>Magento\Config\Block\System\Config\Form\Fieldset</frontend_model>
+        <comment><![CDATA[<p>Online Installment payment it’s design as optional for Installment where the Card Issuer and Acquiring Bank is the same entity (e.g: BNI Card and BNI Acquiring bank).</p>]]></comment>
```
Fix:
> Optional - Enabling this will display additional payment options to customer, for online installment payment where the Card Issuer and Acquiring Bank is the same entity (e.g: BNI Card and BNI Acquiring bank).

Xaxxis replied:
done

rizdaprasetya commented on etc/adminhtml/system/midtrans_snap_offline_payment.xml:
```diff
@@ -0,0 +1,89 @@
+    <group id="offline" translate="label" type="text" sortOrder="60" showInDefault="1" showInWebsite="1" showInStore="1">
+        <label><![CDATA[Optional - Snap Offline Installment Payment Integration]]></label>
+        <frontend_model>Magento\Config\Block\System\Config\Form\Fieldset</frontend_model>
+        <comment><![CDATA[<p>Offline payment it’s design as optional for Installment where the Card Issuer and Acquiring Bank don't have to be same entity (e.g: BNI Card and Mandiri Acquiring Bank)</p>]]></comment>
+        <field id="active" translate="label" type="select" sortOrder="10" showInDefault="1" showInWebsite="1" showInStore="0">
```
Fix:
> Optional - Enabling this will display additional payment options to customer, for offline Installment where the Card Issuer and Acquiring Bank don't have to be same entity (e.g: BNI Card and Mandiri Acquiring Bank)

Xaxxis replied:
done

rizdaprasetya commented on etc/adminhtml/system/midtrans_snap_specific_payment.xml:
```diff
@@ -0,0 +1,85 @@
+    <group id="specific" translate="label" type="text" sortOrder="30" showInDefault="1" showInWebsite="1" showInStore="1">
+        <label><![CDATA[Optional - Snap Specific Payment Integration]]></label>
+        <frontend_model>Magento\Config\Block\System\Config\Form\Fieldset</frontend_model>
+        <comment><![CDATA[<p>Specific payment it’s design as optional for specific payment that specify in Allow payment method field, Midtrans Snap will show directly payment method on the Snap screen.</p>]]></comment>
```
Fix:
> Optional - Enabling this will display additional payment options to customer, for specific payment that are specified in the "Allowed Payment Method" field, Midtrans Snap will show only the listed payment method on the Snap screen.

Xaxxis replied:
done

rizdaprasetya commented on README.md:
```diff
@@ -1,56 +1,123 @@
+    
+    This is the default Snap for Midtrans Magento plugins, Snap payment will auto-enabled when install the Midtrans plugins. Midtrans will show the available payment method on the Snap payment screen.
+
+2. **Specific Payment integration | Optional** 
+    
+    Specific payment it’s design as optional for specific payment that specify in Allow payment method field, Midtrans Snap will show directly payment method on the Snap screen.
+
+3. **Online Installment payment integration | Optional**
+
+    Online Installment payment it’s design as optional for Installment where the Card Issuer and Acquiring Bank is the same entity (e.g: BNI Card and BNI Acquiring bank).
+
+4. **Offline Installment payment integration | Optional**
+
+    Offline payment it’s desing as optional for Installment where the Card Issuer and Acquiring Bank don't have to be same entity (e.g: BNI Card and Mandiri Acquiring Bank)
```
Change according to suggested fix on commented-review of  `midtrans_snap_installment_payment.xml` file

Xaxxis replied:
done

rizdaprasetya commented on README.md:
```diff
@@ -61,6 +128,30 @@ php bin/magento cache:clean
+    *   **Refund:** This option will request refund Online to Midtrans, Midtrans automatically send refund notification and changes order status to **Closed** from notification.
+    *   **Refund Offline**: An offline refund does not trigger request refund to midtrans it’s only refund in Magento side. You need to take action and carry out the refund manually from Midtrans dashboard.After a refund operation, the order status changes to **Closed**. This order status change is controlled by the Magento system. 
+    
+    The order status change does not mean that the refund was carried out successfully on Midtrans side. If the transaction status in Midtrans dashboard changes to **REFUND**, then the refund went through successfully
```
Fix to: 
> The status change may not mean that the refund has carried out successfully on Midtrans side. When the transaction status in Midtrans dashboard changes to **REFUND**, then the refund went through successfully

Xaxxis replied:
done

rizdaprasetya replied:
What's `done`?
I asked to revise the sentence to 
> The status change may not mean that the refund has carried out successfully on Midtrans side. When the transaction status in Midtrans dashboard changes to REFUND, then the refund went through successfully

Why didn't it get revised? And instead you add new sentence with another grammar mistakes.

Xaxxis replied:
should be in accordance with the request @rizdaprasetya , sorry missed for that

Commit:
Some improvement for compability version

Runing well in Magento 2.1.0, 2.20, 2.3.4
Tested with docker

Commit:
Update readme info refund

Commit:
Update docs review

Commit:
Update review 2 for admin dashboard details

Commit:
Merge branch 'master' of https://github.com/Xaxxis/Midtrans-Magento2

Commit:
Update Requirements

rizdaprasetya commented:
There are some review that are still not resolved, please view the full comment from here:
https://github.com/Midtrans/Midtrans-Magento2/pull/10/files#diff-04c6e90faac2675aa89e2176d2eec7d8

Under the file section `readme.md` @Xaxxis

Commit:
Update readme review

rizdaprasetya commented on README.md:
```diff
@@ -1,56 +1,122 @@
-5. Login to your Magento Admin Panel.
-6. In your Magento admin area, enable the Midtrans plug-in and insert your merchant details (Server key and client key) in the Menu "Stores" > "Configuration" > "Sales" > "Payment Method" > Tab "Midtrans Snap".
 
 ### Midtrans&nbsp;  MAP Configuration
```
Please add Snap preference settings, like
https://github.com/veritrans/SNAP-Woocommerce/#midtrans-map-configuration

rizdaprasetya reviewed (approved):
One more minor things on Readme.
After that approved, please merge and release 👍 
Thanks @Xaxxis

Commit:
Update snap pref setting

Commit:
Minimize js to xml
